Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/PolDiffILLReduction.py

```diff
@@ -0,0 +1,876 @@
+            if 'container_formula' in self._experimentProperties:
+                kwargs['ContainerRadius'] = self._experimentProperties['container_radius'].value
+        elif geometry_type == 'Annulus':
+            kwargs['SampleInnerRadius'] = self._experimentProperties['inner_radius'].value
+            kwargs['SampleOuterRadius']  = self._experimentProperties['outer_radius'].value
+            if 'container_formula' in self._experimentProperties:
+                kwargs['ContainerInnerRadius'] = self._experimentProperties['container_inner_radius'].value
+                kwargs['ContainerOuterRadius'] = self._experimentProperties['container_outer_radius'].value
+
+        self._enforce_uniform_units(sample_ws, container_ws)
+        if geometry_type == 'Custom':
+            raise RuntimeError('Custom geometry treatment has not been implemented yet.')
+        else:
+            for entry_no, entry in enumerate(mtd[sample_ws]):
+                correction_ws = '{}_corr'.format(geometry_type)
+                if ( (self._method_data_structure == 'Uniaxial' and entry_no % 2 == 0)
+                     or (self._method_data_structure == 'XYZ' and entry_no % 6 == 0)
+                     or (self._method_data_structure == '10p' and entry_no % 10 == 0) ):
+                    PaalmanPingsMonteCarloAbsorption(SampleWorkspace=entry,
+                                                     Shape=geometry_type,
+                                                     CorrectionsWorkspace=correction_ws,
+                                                     ContainerWorkspace=mtd[container_ws].getItem(entry_no),
+                                                     **kwargs)
+                ApplyPaalmanPingsCorrection(SampleWorkspace=entry,
+                                            CorrectionsWorkspace=correction_ws,
+                                            CanWorkspace=mtd[container_ws].getItem(entry_no),
+                                            OutputWorkspace=entry)
+
+        return sample_ws
+
+    def _component_separation(self, ws):
+        """Separates coherent, incoherent, and magnetic components based on spin-flip and non-spin-flip intensities of the
+        current sample. The method used is based on either the user's choice or the provided data structure."""
+
+        nMeasurements = 0
+        nComponents = 0
+        componentNames = ['Coherent', 'Incoherent', 'Magnetic']
+        if self._user_method == '10p':
+            nMeasurements = 10
+            nComponents = 3
+        elif self._user_method == 'XYZ':
+            nMeasurements = 6
+            nComponents = 3
+        elif self._user_method == 'Uniaxial':
+            nMeasurements = 2
+            nComponents = 2
+
+        tmp_names = []
+        for entry_no in range(0, mtd[ws].getNumberOfEntries(), nMeasurements):
+            dataY_nuclear = np.zeros(shape=(mtd[ws].getItem(entry_no).getNumberHistograms(), mtd[ws].getItem(entry_no).blocksize()))
```

Comment:
you should get the block size and the number of histograms only once and the reuse them